add a probe_use_gitignore option

campbellr opened the pull request:
This lets you automatically ignore any files or directories in
.gitignore (or any other input files to the git exclude mechanism).

torbiak commented:
Thanks, Ryan! I have some thoughts:

git-check-ignore errors when used outside of a repository, and I'm concerned about the performance of spawning a process for each scanned file, especially on cygwin where forks are hella slow and have a decent chance of failing. git-ls-files could provide the whole list in one shot, with git ignore rules applied if desired.

torbiak commented on plugin/probe.vim:
```diff
@@ -54,6 +54,10 @@ if !exists('g:probe_mappings')
 endif
 
+if !exists('g:probe_use_gitignore')
+    let g:prove_use_gitignore = 0
```
prove -> probe

campbellr replied:
Oh good catch.

torbiak commented on doc/probe.txt:
```diff
@@ -186,6 +186,13 @@ OPTIONS                                                        *probe-options*
+    Use the current repositories .gitignore file to when scanning for files.
+    Default: (boolean) 0
+
+    g:probe_ignore_files and g:probe_use_wildignore  still function normally
```
extra space before "still"

campbellr commented:
>git-check-ignore errors when used outside of a repository

That's true. I figured that was acceptable since it would just result in the file/directory not being ignored, which is the old behaviour. `find_in_repo` ends up moving to repo root anyways.

Am I missing something?

 > and I'm concerned about the performance of spawning a process for each scanned file, especially on cygwin where forks are hella slow and have a decent chance of failing. git-ls-files could provide the whole list in one shot, with git ignore rules applied if desired.

I was worried about that too, but (on linux at least) it didn't seem to be significantly slower.

What are your thoughts on just adding the result of 'git ls-files --others --ignored --exclude-standard` into the ignore list? is that what you were referring to above?

torbiak commented:
> Am I missing something?
I was wondering if rules from `~/.gitignore` could be used outside of repos, which could make it worth running git-check-ignore on files outside of repos, but it doesn't seem like it.

On cygwin, scanning 40k files under `/usr` takes 16s without running `git-check-ignore`, and a very long time with it. [It's still running maybe 15 minutes later.]

I was thinking of not using scan_files if we're in a repo, instead getting the whole list from git-ls-files, and then filtering it using g:probe_ignore_files.

campbellr commented:
@torbiak I made some these changes before seeing your latest comment. I'll look into just using `git ls-files`.

campbellr committed:
fix find_metadir

In the fix for #9, `metadir_pattern` was changed to include a newline, but
that newline was never stripped off before returning it.

So instead of `s:find_metadir` returning something like `path/to/.git`, it
returned `path/to/.git\n` which broke the check in `s:branch`

campbellr commented:
@torbiak The latest changes collect the file list using `git ls-files` and then filter with `g:probe_ignore_files`. I also fixed a minor bug in `s:find_metadir` that was breaking `s:branch`.

I'm no longer sure that the name of the `g:probe_use_gitignore` option is entirely appropriate anymore though... Thoughts?

torbiak commented:
ah, I just realized that `git ls-files` doesn't enumerate files in submodules and I haven't found a way to include them. Doing a `git submodules foreach git ls-files -co` takes a surprisingly long time: for my dotfiles dir it takes 2s on cygwin, compared to 0.3s to scan using `globpath()` in probe.

I'm worried that copying git's ignored files into the probe's list of ignore patterns will perform poorly if there are many ignored files, as there can often be. Adding 200 simple patterns to `g:probe_ignore_files` increased the duration of the 40k file scan of `/usr` from 16s to 89s. Maybe we can have probe scan as before, but then filter the whole list at once with something like `system('git check-ignore --stdin --non-matching --verbose', filepaths)`.

> @torbiak The latest changes collect the file list using git ls-files and then filter with g:probe_ignore_files. I also fixed a minor bug in s:find_metadir that was breaking s:branch.

ah, that's why I changed `metadir_pattern` to look for a *zero-length* word boundary instead. I thought about that for a while and still botched it. I was only thinking about what was being matched, not what was being captured.

> I'm no longer sure that the name of the g:probe_use_gitignore option is entirely appropriate anymore though... Thoughts?

I still like `g:probe_use_gitignore` as it describes the end result for the user.

campbellr commented:
Hm, good catch with the submodules, i'll look into that.

> Doing a git submodules foreach git ls-files -co takes a surprisingly long time:

Wow, windows/cygwin performance must be terrible, because on my machine running `git ls-files; git submodule foreach -q 'for file in $(git ls-files); do echo "$path/$file"; done'` on your dotfiles takes 0.087s while `globpath()` takes 0.05s.

campbellr committed:
add a probe_use_gitignore option

This lets you automatically ignore any files or directories in
.gitignore (or any other input files to the git exclude mechanism).

campbellr commented:
@torbiak 

OK, i finally got around to looking at this again.

Unfortunately the `git check-ignore --stdin` is slow **and** doesn't handle submodules, so i decide to just update my original solution to use `git ls-files` with `git submodule foreach`.

It seems to work fairly well and while it is a bit slower for smaller repos (with lots of submodules?), it seems to be significantly faster for large repos.

Some numbers from my laptop (2010 macbook pro w/o SSD):

Repo | scan_files | scan_git
-------- | ------------- | -----------
**torbiak/dotfiles** | 0.17s | 0.61s 
**torvalds/linux** | 10.33s | 2.31s

torbiak commented:
> It seems to work fairly well and while it is a bit slower for smaller repos (with lots of submodules?), it seems to be significantly faster for large repos.

Seems like the right tradeoff.

torbiak commented on autoload/probe/file.vim:
```diff
@@ -72,6 +76,26 @@ function! probe#file#scan()
+    let files = []
+    echo "Scanning " . s:find_repo_root()
+
+    let cmd = "git ls-files; git submodule foreach -q 'for file in $(git ls-files); do echo \"$path/$file\"; done'"
```
Include `--other` files?

campbellr replied:
The `--other` flag makes `git ls-files` return the opposite of what I want (untracked and ignored files).

I assume you are suggesting that option because you want untracked (but not ignored) files to show up? 

I actually like excluding untracked files but it probably be nice to make that configurable.

torbiak replied:
Having both `--cached` and `--other` is closest the `globpath` scan behaviour, returning all files that aren't explicitly ignored, so I was assuming that was the way to go.

hehe, if I recall correctly your workflows result in almost never have untracked files in your repo, but I speak for those undisciplined among us who do.

campbellr replied:
Ah OK, it turns out that I just misunderstood how the flags in `git ls-files` actually worked.

It seems like `git ls-files --cached --others --exclude-standard` works fine (although I might need to add another option to ignore untracked at some point...).

torbiak commented on doc/probe.txt:
```diff
@@ -186,6 +186,13 @@ OPTIONS                                                        *probe-options*
     g:probe_ignore_files still functions normally when this option is set.
 
+|g:probe_use_gitignore|                              *g:probe_use_gitignore*
+    Use the current repositories .gitignore file to when scanning for files.
```
More accurately we're using git's "ignore settings", but I don't know what the right name for them is. The global and `info/exclude` files are taken into account, too, right?

campbellr replied:
Yes. I could reword that to something like:

> Ignore any file excluded by .gitignore (or other relevant exclude files) when scanning for files.

torbiak replied:
Nice.

campbellr commented:
One other difference with the change to use `git ls-files` that is worth mentioning is that hidden files/directories (eg: .gitignore) show up now.

Not sure if that's universally a good thing.

torbiak commented:
I think it's good. "Show me everything I haven't explicitly hidden."

torbiak commented:
Works great for me. Thanks for sticking with this. It was trickier than I expected.

torbiak commented:
> One other difference with the change to use git ls-files that is worth mentioning is that hidden files/directories (eg: .gitignore) show up now.

oh, I don't think I ever realized that the `globpath` scanner excluded them.

campbellr commented:
Awesome! 

And thanks for all the helpful comments.
